File: packages/automationstudio-types-node/src/adapter_registry.ts
import { ActionDefinition } from './generated/execution';
import { AdapterContract } from './generated/adapter';

export class AdapterConflictError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'AdapterConflictError';
  }
}

export class AdapterNotFoundError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'AdapterNotFoundError';
  }
}

export class AdapterRegistry {
  private readonly adapters: AdapterContract[] = [];

  register(adapter: AdapterContract): void {
    this.adapters.push(adapter);
  }

  selectAdapter(action: ActionDefinition): AdapterContract {
    const candidates = this.adapters.filter(adapter => 
      adapter.supportedActionTypes.includes(action.type)
    );

    if (candidates.length === 0) {
      throw new AdapterNotFoundError(`No adapter found for action type '${action.type}'`);
    }

    const maxPriority = Math.max(...candidates.map(a => a.priority));
    const topCandidates = candidates.filter(a => a.priority === maxPriority);

    if (topCandidates.length > 1) {
      const ids = topCandidates.map(a => a.id).join(', ');
      throw new AdapterConflictError(
        `Multiple adapters with priority ${maxPriority} matched action '${action.type}': [${ids}]. ` +
        `Adapter selection must be deterministic.`
      );
    }

    const selected = topCandidates[0];
    if (!selected) {
      // Keep the contract total even if the selection logic changes later.
      throw new AdapterNotFoundError(`No adapter found for action type '${action.type}'`);
    }

    return selected;
  }
}

File: packages/automationstudio-sdk/src/automationstudio/sdk/execution/adapter_registry.py
from typing import List, Dict, Optional, Any
from automationstudio.sdk.models.generated.adapter import AdapterContract
from automationstudio.sdk.models.generated.execution import ActionDefinition

class AdapterConflictError(Exception):
    pass

class AdapterNotFoundError(Exception):
    pass

class AdapterRegistry:
    def __init__(self):
        self._adapters: List[AdapterContract] = []

    def register(self, adapter: AdapterContract):
        self._adapters.append(adapter)

    def select_adapter(self, action: ActionDefinition) -> AdapterContract:
        candidates = []
        for adapter in self._adapters:
            # simple capability/support check based on the schema
            action_type_str = action.type.value
            if action_type_str in adapter.supportedActionTypes:
                # Assuming target object evaluation can be more complex, but for now we just check if it claims support
                candidates.append(adapter)

        if not candidates:
            raise AdapterNotFoundError(f"No adapter found for action type '{action.type.value}'")

        # Select by highest priority (assuming higher number is higher priority)
        max_priority = max(adapter.priority for adapter in candidates)
        top_candidates = [a for a in candidates if a.priority == max_priority]

        if len(top_candidates) > 1:
            ids = [a.id for a in top_candidates]
            raise AdapterConflictError(
                f"Multiple adapters with priority {max_priority} matched action '{action.type.value}': {ids}. "
                f"Adapter selection must be deterministic."
            )

        return top_candidates[0]
